node/src/proxy_server/mod.rs

```diff
@@ -4934,6 +5879,97 @@ mod tests {
+        let main_cryptde = main_cryptde();
+        let alias_cryptde = alias_cryptde();
+        let http_request = b"GET /index.html HTTP/1.1\r\nHost: nowhere.com\r\n\r\n";
+        let (neighborhood_mock, _, _) = make_recorder();
+        let destination_key = PublicKey::from(&b"our destination"[..]);
+        let neighborhood_mock = neighborhood_mock.route_query_response(Some(RouteQueryResponse {
+            route: Route { hops: vec![] },
+            expected_services: ExpectedServices::RoundTrip(
+                vec![make_exit_service_from_key(destination_key.clone())],
+                vec![],
+                1234,
+            ),
+        }));
+        let socket_addr = SocketAddr::from_str("1.2.3.4:5678").unwrap();
+        let stream_key = make_meaningless_stream_key();
+        let expected_data = http_request.to_vec();
+        let msg_from_dispatcher = InboundClientData {
+            timestamp: SystemTime::now(),
+            peer_addr: socket_addr.clone(),
+            reception_port: Some(HTTP_PORT),
+            sequence_number: Some(0),
+            last_data: true,
+            is_clandestine: false,
+            data: expected_data.clone(),
+        };
+        let expected_payload = ClientRequestPayloadFactoryReal::new()
+            .make(
+                &msg_from_dispatcher,
+                stream_key.clone(),
+                alias_cryptde,
+                &Logger::new("test"),
+            )
+            .unwrap();
+        let stream_key_factory = StreamKeyFactoryMock::new().make_result(stream_key.clone());
+        let system = System::new(
+            "proxy_server_receives_http_request_from_dispatcher_then_sends_cores_package_to_hopper",
+        );
+        let mut subject = ProxyServer::new(
+            main_cryptde,
+            alias_cryptde,
+            true,
+            Some(STANDARD_CONSUMING_WALLET_BALANCE),
+            false,
+        );
+        subject.stream_key_factory = Box::new(stream_key_factory);
+        let subject_addr: Addr<ProxyServer> = subject.start();
+        let mut peer_actors = peer_actors_builder()
+            .neighborhood(neighborhood_mock)
+            .build();
+        peer_actors.proxy_server = ProxyServer::make_subs_from(&subject_addr);
+        subject_addr.try_send(BindMessage { peer_actors }).unwrap();
+
+        subject_addr.try_send(msg_from_dispatcher).unwrap();
+
+        subject_addr
+            .try_send(AssertionsMessage {
+                assertions: Box::new(move |proxy_server: &mut ProxyServer| {
+                    let dns_retry = proxy_server.dns_failure_retries.get(&stream_key).unwrap();
+                    assert_eq!(dns_retry.retries_left, 3);
+                    assert_eq!(dns_retry.unsuccessful_request, expected_payload);
+                }),
+            })
+            .unwrap();
+        System::current().stop();
+        system.run();
+    }
+
+    #[test]
+    fn hostname_works() {
```

Comment:
I think I'd like to see an example with just `http` (not `https`), and I'd like to see the really wild examples cause errors, rather than returning host names everybody knows are wrong.

Reply:
I've added a new card for this issue. and added your comments.
https://github.com/MASQ-Project/MASQ-Node-issues/issues/743